pkg/streamingpromql/optimize/plan/eliminate_deduplicate_and_merge_test.go

```diff
@@ -238,23 +501,106 @@ func TestEliminateDeduplicateAndMergeOptimizationPassPlan(t *testing.T) {
+					- FunctionCall: abs(...)
+						- DeduplicateAndMerge
+							- FunctionCall: ceil(...)
+								- DeduplicateAndMerge
+									- FunctionCall: label_replace(...)
+										- param 0: DeduplicateAndMerge
+											- FunctionCall: label_replace(...)
+												- param 0: FunctionCall: rate(...)
+													- MatrixSelector: {__name__="foo"}[5m0s]
+												- param 1: StringLiteral: "dst1"
+												- param 2: StringLiteral: "$1"
+												- param 3: StringLiteral: "src1"
+												- param 4: StringLiteral: "(.*)"
+										- param 1: StringLiteral: "dst2"
+										- param 2: StringLiteral: "$1"
+										- param 3: StringLiteral: "dst1"
+										- param 4: StringLiteral: "(.*)"
+				`,
+			// keeps:
+			// - DeduplicateAndMerge around  both label_replace - to deduplicate after label_replace might modify labels and therefore potentially introduce duplicates.
+			// drops:
+			// - DeduplicateAndMerge wrapping rate is dropped because selector has exact name matcher.
+			// - DeduplicateAndMerge wrapping ceil function is dropped because name removal is delayed and series will be deduplicated later.
+			// - DeduplicateAndMerge at the root is dropped, because selector has exact name matcher. label_replace can't reintroduce __name__ label because it's dropped after. Even if it will modify it, DeduplicateAndMerge right after will hanlde this.
+			expectedPlanWithDelayedNameRemoval: `
+			- DropName
+				- FunctionCall: abs(...)
+					- FunctionCall: ceil(...)
+						- DeduplicateAndMerge
+							- FunctionCall: label_replace(...)
+								- param 0: DeduplicateAndMerge
+									- FunctionCall: label_replace(...)
+										- param 0: FunctionCall: rate(...)
+											- MatrixSelector: {__name__="foo"}[5m0s]
+										- param 1: StringLiteral: "dst1"
+										- param 2: StringLiteral: "$1"
+										- param 3: StringLiteral: "src1"
+										- param 4: StringLiteral: "(.*)"
+								- param 1: StringLiteral: "dst2"
+								- param 2: StringLiteral: "$1"
+								- param 3: StringLiteral: "dst1"
+								- param 4: StringLiteral: "(.*)"
+			`,
+			nodesEliminatedWithoutDelayed: 2,
+			nodesEliminatedWithDelayed:    3,
 		},
 
 		// Test cases to confirm we're skipping optimization when expression has a binary operation.
 		"or operator - should keep DeduplicateAndMerge": {
 			expr: `foo or bar`,
-			expectedPlan: `
+			expectedPlanWithoutDelayedName: `
 				- DeduplicateAndMerge
 					- BinaryExpression: LHS or RHS
 						- LHS: VectorSelector: {__name__="foo"}
 						- RHS: VectorSelector: {__name__="bar"}
 				`,
-			nodesEliminated: 0,
+			expectedPlanWithDelayedNameRemoval: `
+				- DeduplicateAndMerge
+					- DropName
+						- BinaryExpression: LHS or RHS
+							- LHS: VectorSelector: {__name__="foo"}
+							- RHS: VectorSelector: {__name__="bar"}
+				`,
+			nodesEliminatedWithoutDelayed: 0,
+			nodesEliminatedWithDelayed:    0,
 		},
 		"or operator with functions - should keep DeduplicateAndMerge": {
```

Comment:
Might be worth adding a note here that we could remove the `DeduplicateAndMerge` nodes on each side, but we currently don't for simplicity.